### PR TITLE
[LLDB][SBSaveCore] Add Extension to Save a thread and N pointers deep

### PR DESCRIPTION
**Pull request by Jlalond**
This adds an extension to allow the saving of a Thread and it's pointers N layers deep via BFS. This lets SBSaveCore 'automagically' figure out everything important for a given thread to save on your behalf.

### CODE REVIEW

File: lldb/test/API/functionalities/process_save_core_minidump/TestProcessSaveCoreMinidump.py [hunk 1]
```diff
@@ -470,6 +471,27 @@ def save_core_with_region(self, process, region_index):
             if os.path.isfile(custom_file):
                 os.unlink(custom_file)
 
+    def save_core_one_thread_one_heap(self, process, region_index):
```
Comment (jeffreytan81) [inline lldb/test/API/functionalities/process_save_core_minidump/TestProcessSaveCoreMinidump.py:474]: Who is invoking this method? 
Is `region_index` used?

File: lldb/bindings/interface/SBSaveCoreOptionsExtensions.i [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+                if var.TypeIsPointerType():
+                    queue.append(var.Dereference())
+            
+            while (num_pointers_deep > 0 and len(queue) > 0):
+                queue_size = len(queue)
+                for i in range(0, queue_size):
+                    var = queue.pop(0)
+                    memory_region = lldb.SBMemoryRegionInfo()
+                    process.GetMemoryRegionInfo(var.GetAddress().GetOffset(), memory_region)
+                    self.AddMemoryRegionToSave(memory_region)
```
Comment (jeffreytan81) [inline lldb/bindings/interface/SBSaveCoreOptionsExtensions.i:20]: What if the pointer points to inaccessible memory or the pointer is corrupted? You need to check the validity of the memory region?

File: lldb/bindings/interface/SBSaveCoreOptionsExtensions.i [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+#ifdef SWIGPYTHON
+    %pythoncode% {
+        '''Add a thread to the SaveCoreOptions thread list, and follow it's children N pointers deep, adding each memory region to the SaveCoreOptions Memory region list.'''
+        def save_thread_with_heaps(self, thread, num_pointers_deep = 3):
+            self.AddThread(thread)
+            frame = thread.GetFrameAtIndex(0)
+            process = thread.GetProcess()
+            queue = []
+            for var in frame.locals:
+                if var.TypeIsPointerType():
```
Comment (jeffreytan81) [inline lldb/bindings/interface/SBSaveCoreOptionsExtensions.i:11]: This API is deprecated per doc. You should use SBType::IsPointerType() and also check reference type.

File: lldb/bindings/interface/SBSaveCoreOptionsExtensions.i [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+                    queue.append(var.Dereference())
+            
+            while (num_pointers_deep > 0 and len(queue) > 0):
+                queue_size = len(queue)
+                for i in range(0, queue_size):
+                    var = queue.pop(0)
+                    memory_region = lldb.SBMemoryRegionInfo()
+                    process.GetMemoryRegionInfo(var.GetAddress().GetOffset(), memory_region)
```
Comment (jeffreytan81) [inline lldb/bindings/interface/SBSaveCoreOptionsExtensions.i:19]: Do you need to perform memory region merging? For example, if two pointers point to the same objects or different parts of the same memory region. Is this a problem?

**Review by jeffreytan81 — CHANGES_REQUESTED** [review]
I think we can do better than save at memory region fidelity which is at least at page boundary if I am not wrong. 

We can use type information to query and save the object's size without including hundreds of unrelated objects in the heap happen to be in the same page. This should greatly reduce the size of the dump.

File: lldb/bindings/interface/SBSaveCoreOptionsExtensions.i (hunk 3 above)
Comment (Jlalond) [reply lldb/bindings/interface/SBSaveCoreOptionsExtensions.i:11]: Oh, I didn't know that. Thanks!

File: lldb/bindings/interface/SBSaveCoreOptionsExtensions.i (hunk 4 above)
Comment (Jlalond) [reply lldb/bindings/interface/SBSaveCoreOptionsExtensions.i:19]: The new RangeMap will deduplicate memory based on permissions.

File: lldb/bindings/interface/SBSaveCoreOptionsExtensions.i (hunk 2 above)
Comment (Jlalond) [reply lldb/bindings/interface/SBSaveCoreOptionsExtensions.i:20]: The save-core impl will actually just omit a non readable/corrupted memory.

We could have the concern that we get an a memory range that is massive, like `2^60-2^64`, but I think that's a generic concern of this API

**Comment by Jlalond** [comment]
> I think we can do better than save at memory region fidelity which is at least at page boundary if I am not wrong.

> We can use type information to query and save the object's size without including hundreds of unrelated objects in the 
> heap happen to be in the same page. This should greatly reduce the size of the dump.

I'm okay with this, when @clayborg and I talked about this the original assumption is we should do the entire memory region it's in, because with deduplication we will likely get multiple pointers to the same heap. I think we should leave this heap-region mode, and then also build a 'just save my objects' extension.

**Comment by jeffreytan81** [comment]
> > I think we can do better than save at memory region fidelity which is at least at page boundary if I am not wrong.
> 
> > We can use type information to query and save the object's size without including hundreds of unrelated objects in the
> > heap happen to be in the same page. This should greatly reduce the size of the dump.
> 
> I'm okay with this, when @clayborg and I talked about this the original assumption is we should do the entire memory region it's in, because with deduplication we will likely get multiple pointers to the same heap. I think we should leave this heap-region mode, and then also build a 'just save my objects' extension.

The value of this option lies in reducing the size to a minimum while retaining useful information. In such cases, saving object closures is a superior choice compared to saving memory regions. You can easily achieve minidumps that are 2 to 3 times smaller in some cases.

**Comment by Jlalond** [comment]
> > > I think we can do better than save at memory region fidelity which is at least at page boundary if I am not wrong.
> > 
> > 
> > > We can use type information to query and save the object's size without including hundreds of unrelated objects in the
> > > heap happen to be in the same page. This should greatly reduce the size of the dump.
> > 
> > 
> > I'm okay with this, when @clayborg and I talked about this the original assumption is we should do the entire memory region it's in, because with deduplication we will likely get multiple pointers to the same heap. I think we should leave this heap-region mode, and then also build a 'just save my objects' extension.
> 
> The value of this option lies in reducing the size to a minimum while retaining useful information. In such cases, saving object closures is a superior choice compared to saving memory regions. You can easily achieve minidumps that are 2 to 3 times smaller in some cases.

Switched to a per object size.

**Comment by Jlalond** [comment]
This does require us to walk the stack backwards instead of just taking the `0th` frame

**Pull request closed without merging**